crates/sui-core/src/execution_cache/unit_tests/writeback_cache_tests.rs

```diff
@@ -0,0 +1,515 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use rand::{rngs::StdRng, SeedableRng};
+use std::{
+    collections::BTreeMap,
+    future::Future,
+    sync::atomic::Ordering,
+    sync::{atomic::AtomicU32, Arc},
+};
+use sui_test_transaction_builder::TestTransactionBuilder;
+use sui_types::{
+    base_types::{random_object_ref, SuiAddress},
+    crypto::{deterministic_random_account_key, get_key_pair_from_rng, AccountKeyPair},
+    object::{MoveObject, Owner, OBJECT_START_VERSION},
+};
+use tempfile::tempdir;
+
+use super::*;
+use crate::{
+    authority::{authority_store_tables::AuthorityPerpetualTables, AuthorityStore},
+    execution_cache::ExecutionCacheAPI,
+    test_utils::init_state_parameters_from_rng,
+};
+
+async fn init_authority_store() -> Arc<AuthorityStore> {
+    let seed = [1u8; 32];
+    let (genesis, _) = init_state_parameters_from_rng(&mut StdRng::from_seed(seed));
+    let committee = genesis.committee().unwrap();
+
+    // Create a random directory to store the DB
+    let dir = tempdir().unwrap();
+    let db_path = dir.path();
+
+    let perpetual_tables = Arc::new(AuthorityPerpetualTables::open(db_path, None));
+    AuthorityStore::open_with_committee_for_testing(perpetual_tables, &committee, &genesis, 0)
+        .await
+        .unwrap()
+}
+
+trait AssertInserted {
+    fn assert_inserted(&self);
+}
+
+impl<T> AssertInserted for Option<T> {
+    fn assert_inserted(&self) {
+        assert!(self.is_none());
+    }
+}
+
+impl AssertInserted for bool {
+    fn assert_inserted(&self) {
+        assert!(*self);
+    }
+}
+
+type ActionCb = Box<dyn Fn(&mut Scenario)>;
+
+struct Scenario {
+    store: Arc<AuthorityStore>,
+    cache: Box<WritebackCache>,
+
+    id_map: BTreeMap<u32, ObjectID>,
+    objects: BTreeMap<ObjectID, Object>,
+    outputs: TransactionOutputs,
+    transactions: BTreeSet<TransactionDigest>,
+
+    action_count: Arc<AtomicU32>,
+    do_after: Option<(u32, ActionCb)>,
+}
+
+impl Scenario {
+    async fn new(do_after: Option<(u32, ActionCb)>, action_count: Arc<AtomicU32>) -> Self {
+        let store = init_authority_store().await;
+        let cache = Box::new(WritebackCache::new_with_no_metrics(store.clone()));
+        Self {
+            store,
+            cache,
+            id_map: BTreeMap::new(),
+            objects: BTreeMap::new(),
+            outputs: Self::new_outputs(),
+            transactions: BTreeSet::new(),
+
+            action_count,
+            do_after,
+        }
+    }
+
+    fn cache(&self) -> &dyn ExecutionCacheAPI {
+        &*self.cache
+    }
+
+    fn count_action(&mut self) {
+        let prev = self.action_count.fetch_add(1, Ordering::Relaxed);
+        if let Some((count, f)) = &self.do_after.take() {
```

Comment:
I'm still learning this. Won't the first `take()` move the `ActionCb` out of the do_after? Should the `take()` happen after `prev == *count`?

Reply:
you are of course correct and i have no idea why i wrote that. the test output seemed to indicate everything was working but i obviously didn't look closely enough